### PR TITLE
Add copy function for DxilSignature and DxilTypeSystem.

### PR DESCRIPTION
**Pull request by python3kgae**
Add copy function for DxilSignature and DxilTypeSystem.
Also make OP::RefreshCache() public.

### CODE REVIEW

File: lib/HLSL/DxilTypeSystem.cpp [hunk 1]
```diff
@@ -253,6 +253,55 @@ StructType *DxilTypeSystem::GetNormFloatType(CompType CT, unsigned NumComps) {
...
+        CopyTypeAnnotation(Ty, src);
+      }
+    }
+  }
+}
+
+void DxilTypeSystem::CopyFunctionAnnotation(const llvm::Function *pDstFunction,
+                              const llvm::Function *pSrcFunction,
+                              DxilTypeSystem &src) {
+  DxilFunctionAnnotation *annot = src.AddFunctionAnnotation(pSrcFunction);
```
Comment (dmpots) [inline lib/HLSL/DxilTypeSystem.cpp]: Why is this `AddFunctionAnnotation`? It seems like `GetFunctionAnnotation` would be better.

File: include/dxc/HLSL/DxilTypeSystem.h [hunk 2]
```diff
@@ -180,6 +180,12 @@ class DxilTypeSystem {
   llvm::StructType *GetSNormF32Type(unsigned NumComps);
   llvm::StructType *GetUNormF32Type(unsigned NumComps);
 
+  // Methods to copy annotation from another DxilTypeSystem.
+  void CopyTypeAnnotation(const llvm::Type *Ty, DxilTypeSystem &src);
+  void CopyFunctionAnnotation(const llvm::Function *pDstFunction,
+                              const llvm::Function *pSrcFunction,
+                              DxilTypeSystem &src);
```
Comment (dmpots) [inline include/dxc/HLSL/DxilTypeSystem.h]: The `src` param should be const. We should not modify the input type system.

File: lib/HLSL/DxilSignature.cpp [hunk 3]
```diff
@@ -43,6 +43,19 @@ unique_ptr<DxilSignatureElement> DxilSignature::CreateElement() {
   return unique_ptr<DxilSignatureElement>(new DxilSignatureElement(m_sigPointKind));
 }
 
+void DxilSignature::CopySignatureElements(DxilSignature &src) {
+  const bool bSetID = false;
+  for (auto &Elt : src.GetElements()) {
+    std::unique_ptr<DxilSignatureElement> newElt = CreateElement();
```
Comment (dmpots) [inline lib/HLSL/DxilSignature.cpp]: Should we add a copy constructor for `DxilSignatureElement` to simplify this code?

File: lib/HLSL/DxilTypeSystem.cpp [hunk 4]
```diff
@@ -253,6 +253,55 @@ StructType *DxilTypeSystem::GetNormFloatType(CompType CT, unsigned NumComps) {
...
+    if (DxilStructAnnotation *annot = src.GetStructAnnotation(ST)) {
+      DxilStructAnnotation *dstAnnot = AddStructAnnotation(ST);
+      // Copy the annotation.
+      *dstAnnot = *annot;
+      // Copy field type annotations.
+      for (Type *Ty : ST->elements()) {
+        CopyTypeAnnotation(Ty, src);
+      }
+    }
+  }
```
Comment (dmpots) [inline lib/HLSL/DxilTypeSystem.cpp]: What about non-struct types? Do you need to copy those as well?

Comment (python3kgae) [inline lib/HLSL/DxilTypeSystem.cpp]: Only structure type has annotation.

Comment (dmpots) [inline lib/HLSL/DxilTypeSystem.cpp]: Ok, makes sense. In that case an early check and return would make it clear.

    // Only struct type has annotation
    if (!isa<StructType>)
        return;

**Pull request merged**